In [ ]:
#question 1
def sum_even(numbers):
    return sum(num for num in numbers if num % 2 == 0)

In [ ]:
#Question 2
def reverse_string(s):
    return s[::-1]

In [ ]:
#Question 4
def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(n**0.5)+1):
        if n % i == 0:
            return False
    return True

primes = [n for n in range(1, 201) if is_prime(n)]
print(primes)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199]


In [ ]:
#Question 3
def square_list(numbers):
    return [num**2 for num in numbers]

In [ ]:
#Question 5
class Fibonacci:
    def __init__(self, terms):
        self.terms = terms
        self.count = 0
        self.a, self.b = 0, 1

    def __iter__(self):
        return self

    def __next__(self):
        if self.count >= self.terms:
            raise StopIteration
        if self.count == 0:
            self.count += 1
            return 0
        elif self.count == 1:
            self.count += 1
            return 1
        else:
            self.a, self.b = self.b, self.a + self.b
            self.count += 1
            return self.b

In [ ]:
#Question 6
def powers_of_two(exponent):
    for i in range(exponent + 1):
        yield 2 ** i


In [ ]:
#Question 7
def read_file(filename):
    with open(filename, 'r') as file:
        for line in file:
            yield line.strip()


In [ ]:
#Question 8
tuples_list = [(1, 3), (2, 1), (3, 2)]
sorted_list = sorted(tuples_list, key=lambda x: x[1])
print(sorted_list)


[(2, 1), (3, 2), (1, 3)]


In [ ]:
#Question 9
celsius = [0, 20, 30, 40]
fahrenheit = list(map(lambda c: (c * 9/5) + 32, celsius))
print(fahrenheit)


[32.0, 68.0, 86.0, 104.0]


In [ ]:
#Question 10
def remove_vowels(s):
    vowels = 'aeiouAEIOU'
    return ''.join(filter(lambda x: x not in vowels, s))

print(remove_vowels("Hello World"))


Hll Wrld


In [ ]:
#Question 11
orders = [
    ['Order-001', 5, 15.0],
    ['Order-002', 2, 45.0],
    ['Order-003', 1, 80.0],
]

result = list(map(lambda order:
                  (order[0], order[1] * order[2] if order[1] * order[2] >= 100 else order[1] * order[2] + 10),
                  orders))

print(result)


[('Order-001', 85.0), ('Order-002', 100.0), ('Order-003', 90.0)]


In [ ]:
+

1. Difference between a function and a method in Python?
   - A function is a block of code that performs a specific task and is defined using def.
   - A method is a function that is associated with an object (belongs to a class).
   - Example:
       # Function
      def greet():
    return "Hello"

        # Method (attached to a string object)
         "hello".upper()

2. Function arguments and parameters?
   - Parameter: The variable listed inside the parentheses in the function definition.
   - Argument: The value passed to the function when it is called.
Example:
      def add(x, y):  # x and y are parameters
       return x + y
       add(3, 4)  # 3 and 4 are arguments

3. Different ways to define and call a function:
  - Normal function definition
  - Lambda function
  - Function using def inside another function (nested)
Example:
# Normal function
def square(x):
    return x * x

# Lambda function
square_lambda = lambda x: x * x

# Calling
print(square(5))
print(square_lambda(5))

4. Purpose of the return statement?
 - Return is used to send the result from a function back to the caller.
Example:
def multiply(a, b):
    return a * b

5. Iterators vs Iterables?

   - Iterable: Object you can loop over (like list, tuple, string).
   - Iterator: Object that produces next values using __next__().
Example:
iterable = [1, 2, 3]
iterator = iter(iterable)
print(next(iterator))  # 1    

6. Concept of generators?
Generators are functions that yield values one at a time using yield.
Example:
def gen_numbers():
    yield 1
    yield 2
    yield 3

7. Advantages of using generators:
   - Memory efficient (don't store entire sequence in memory).
   - Faster for large datasets.  

8. Lambda function and its use?
   - Anonymous (unnamed) small function.
   - Used when a small function is needed for a short period.
Example:
square = lambda x: x * x
print(square(5))

9.Purpose and usage of map() function?
  - Applies a function to every element in a given iterable.
  Example:
nums = [1, 2, 3, 4]
squared = list(map(lambda x: x**2, nums))
print(squared)    

10. Difference between map(), reduce(), and filter():


   - Function	Purpose
map()	Applies a function to every item
filter()	Filters items based on a condition
reduce()	Reduces a list to a single value
Example:

python
Copy
Edit
from functools import reduce

nums = [1, 2, 3, 4]
mapped = list(map(lambda x: x*2, nums))  # [2, 4, 6, 8]
filtered = list(filter(lambda x: x%2==0, nums))  # [2, 4]
reduced = reduce(lambda x,y: x+y, nums)  # 10

11. Pen & Paper) Internal mechanism for sum operation using reduce
    - List: [47,11,42,13]
    - Steps:
    - Initial: 47 + 11 = 58
58 + 42 = 100
100 + 13 = 113
→ Final sum = 113
